staging/src/k8s.io/client-go/tools/cache/shared_informer_health.go

```diff
@@ -0,0 +1,214 @@
+		minConsistentErrors: 10,
+		consistentErrorsCh:  errorReceiverCh,
+	}
+}
+
+type listWatchHealth struct {
+	lock         sync.Mutex
+	startingTime time.Time
+
+	// minTimeToFail minimum time duration after which we start to fail
+	minTimeToFail time.Duration
+	// minConsistentError minimum number of errors before we start to fail
+	minConsistentErrors int
+	// consistentErrors holds the number of errors seen so far
+	consistentErrors int
+
+	// consistentErrorsCh channel on which we receive List&Watch related errors
+	// it is meant to represent a gauge, that is a number that can go up and down
+	consistentErrorsCh <-chan int
+}
+
+func (lw *listWatchHealth) run(stopCh <-chan struct{}) {
+	//TODO: use wait.Until(...)
+	go lw.runInternal(stopCh)
+}
+
+func (lw *listWatchHealth) runInternal(stopCh <-chan struct{}) {
+	for {
+		select {
+		case <-stopCh:
+			return
+		case seenErrors, ok := <-lw.consistentErrorsCh:
+			if !ok {
+				break
+			}
+			lw.lock.Lock()
+			lw.consistentErrors = seenErrors
+			if seenErrors >= lw.minConsistentErrors && lw.startingTime.IsZero() {
+				lw.startingTime = time.Now()
+			} else if seenErrors < lw.minConsistentErrors && !lw.startingTime.IsZero() {
+				lw.startingTime = time.Time{}
+			}
+			lw.lock.Unlock()
+		}
+	}
+}
+
+func (lw *listWatchHealth) health() (int, bool) {
+	lw.lock.Lock()
+	defer lw.lock.Unlock()
+	if lw.startingTime.IsZero() {
+		return 0, true
+	}
+	now := time.Now()
+	elapsed := now.Sub(lw.startingTime)
+	if elapsed > lw.minTimeToFail {
+		return lw.consistentErrors, false
+	}
+	return 0, true
+}
+
+func newDefaultListWatchLockFreeHealth(errorReceiverCh <-chan error) *listWatchLockFreeHealth {
+	lwlf := &listWatchLockFreeHealth{
+		minTimeToFail:       time.Duration(5 * time.Second),
+		minConsistentErrors: 10,
+		consistentErrorsCh:  errorReceiverCh,
+		errCircularBuffer:   atomic.Value{},
+	}
+
+	lwlf.errCircularBuffer.Store(newErrBuffer(5))
+	return lwlf
+}
+
+type listWatchLockFreeHealth struct {
+	startingTime int64
+
+	// minTimeToFail minimum time duration after which we start to fail
+	minTimeToFail time.Duration
+	// minConsistentError minimum number of errors before we start to fail
+	minConsistentErrors int32
+	// consistentErrors holds the number of errors seen so far
+	consistentErrors int32
+
+	// consistentErrorsCh channel on which we receive List&Watch related errors
+	// it is meant to represent a gauge, that is a number that can go up and down
+	consistentErrorsCh <-chan error
+
+	errCircularBuffer atomic.Value
+}
+
+func (lwlf *listWatchLockFreeHealth) run(stopCh <-chan struct{}) {
+	//TODO: use wait.Until(...)
+	go lwlf.runInternal(stopCh)
+}
+
+func (lwlf *listWatchLockFreeHealth) runInternal(stopCh <-chan struct{}) {
+	for {
+		select {
+		case <-stopCh:
+			return
+		case errRcv, ok := <-lwlf.consistentErrorsCh:
+			if !ok {
+				break
+			}
+			seenErrors := atomic.LoadInt32(&lwlf.consistentErrors)
+			if errRcv == nil {
+				seenErrors = 0
```

Comment:
we could also invalidate `errCircularBuffer` so that we always start fresh. We don't have to if `len(errCircularBuffer) < minConsistentErrors`.